From 08c1c11c6f1102e6d01b7513fff185eed198e0de Mon Sep 17 00:00:00 2001
From: "Maneendra,Pokuri Naga" <pnmaneendra@maxlinear.com>
Date: Mon, 2 Dec 2024 13:53:06 +0530
Subject: [PATCH] WLANRTSYS-86519 Adressing Segfault after VAP Addition on top
 of ML VAP associated radios.

Fixup: WLANRTSYS-60843 MLO changes: AP MLD configuration
---
 src/ap/mxl_hostapd.c | 4 +++-
 src/ap/mxl_mld.c     | 5 +++++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index a37260221..922264bd0 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -2782,8 +2782,10 @@ int mxl_hostapd_deinit_gmld(struct hostapd_data *hapd)
 
 	mld_id = hapd->conf->mld_id;
 	gmld = &(hapd->iface->interfaces->mxl_interfaces.g_ap_mld[mld_id]);
-	if (gmld->num_links == MAX_NUM_LINKS)
+	if (gmld->num_links == MAX_NUM_LINKS) {
+		gmld->num_links--;
 		return 0;
+	}
 
 	if (hapd->conf->mld_id >= MAX_SUPPORTED_MLDS) {
 		wpa_printf(MSG_ERROR, "Invalid MLD id %d. Cannot deinit gmld", hapd->conf->mld_id);
diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 3fe8b102a..a829ea0bb 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -106,6 +106,11 @@ int mxl_update_mld_config_iface(struct hostapd_iface *iface, void *ctx)
 				p_ap_mld = &(iface->interfaces->mxl_interfaces.g_ap_mld[j]);
 				if (p_ap_mld->num_links && !hostapd_mac_comp(p_ap_mld->ap_mld_mac, hapd->conf->mxl_bss_conf.ap_mld_mac)) {
 					if (mxl_is_ap_mld_mac_duplicate(hapd, p_ap_mld)) {
+						if (p_ap_mld->num_links == MAX_NUM_LINKS && (hapd == p_ap_mld->affiliated_links[0] || hapd == p_ap_mld->affiliated_links[1])) {
+							p_ap_mld->affiliated_links[0]->conf->mld_id = j;
+							p_ap_mld->affiliated_links[1]->conf->mld_id = j;
+							break;
+						}
 						wpa_printf(MSG_ERROR, "The duplicate AP MLD mac address("MACSTR") detected !!",
 						  MAC2STR(p_ap_mld->ap_mld_mac));
 						return -1;
-- 
2.43.0

